File: korge-ext-spriter/reference/src/main/java/com/brashmonkey/spriter/Meta.java
package com.brashmonkey.spriter;

public class Meta {
	
	Var[] vars;
	
	static class Var {
		int id;
		String name;
		Value def;
		Key[] keys;
		
		static class Key {
			int id;
			long time;
			Value value;
			
			public Value getValue() {
				return value;
			}
			
			public Class<?> getType() {
				return this.value.getClass();
			}
			
			public int getId() {
				return this.id;
			}
			
			public long getTime() {
				return this.time;
			}
		}
		
		static class Value {
			Object value;
			
			public int getInt() {
				return (Integer)value;
			}
			
			public long getLong() {
				return (Long)value;
			}
			
			public String getString() {
				return (String)value;
			}
		}
		
		public Key get(long time) {
			for (Key key: this.keys)
				if (key.time == time)
					return key;
			return null;
		}
		
		public boolean has(long time) {
			return this.get(time) != null;
		}
		
		public String getName() {
			return this.name;
		}
		
		public int getId() {
			return this.id;
		}
		
		public Value getDefault() {
			return this.def;
		}
	}
	
	public Var getVar(long time) {
		for (Var var: this.vars) 
			if (var.get(time) != null)
				return var;
		return null;
	}
}

File: korge-spriter/src/commonMain/kotlin/com/soywiz/korge/ext/spriter/com/brashmonkey/spriter/Meta.kt
package com.soywiz.korge.ext.spriter.com.brashmonkey.spriter

import kotlin.reflect.*

class Meta {

	var vars: Array<Var>? = null

	class Var {
		var id: Int = 0
		var name: String? = null
		var default: Value? = null
		var keys: Array<Key>? = null

		class Key {
			var id: Int = 0
			var time: Long = 0
			var value: Value? = null

			val type: KClass<*>
				get() = this.value!!::class
		}

		class Value {
			var value: Any? = null

			val int: Int
				get() = (value as Int?)!!

			val long: Long
				get() = (value as Long?)!!

			val string: String
				get() = value as String
		}

		operator fun get(time: Long): Key? {
			for (key in this.keys!!)
				if (key.time == time)
					return key
			return null
		}

		fun has(time: Long): Boolean {
			return this[time] != null
		}
	}

	fun getVar(time: Long): Var? {
		for (`var` in this.vars!!)
			if (`var`[time] != null)
				return `var`
		return null
	}
}
